### DISCUSSION TITLE
Bad result of testing a model by using python interface

### DISCUSSION THREAD

**Discussion by gust-07** [deepmodeling/deepmd-kit#3903] [Q&A] [answered]
I have trained a model, both _lcurve.out_ and _dp test_ shows good quality of this model.  But when I uses python interface, _dp.eval_ to recalculate the energies and forces based on the training data, bad results came out, especially the forces of the last three atoms were much different to the training result.
I have also used the tutorial data  (https://bohrium.dp.tech/notebooks/52879961357) to do the same test, no differences were seen between  the training data and prediction data.
**SO, I doubt any special of my model  or just the input setting is not  right.**

This is the lcurve:
[lcurve.txt](https://github.com/user-attachments/files/15971386/lcurve.txt)

This is how I retest the model:
import numpy as np
import dpdata
from deepmd.infer import DeepPot as DP

#training_systems = dpdata.LabeledSystem("../training_data", fmt = "deepmd/npy")
#predict = training_systems.predict("graph.pb")
#print("orignal energy:%s"%training_systems['energies'])
#print("predicted energy:%s"%predict['energies'])

item = ["coord","box","energy","force"]
data = {}
for i in item:
    data[i] = np.load("../training_data/set.000/%s.npy"%i)

atype = np.loadtxt("../training_data/type.raw")
dp = DP("graph.pb")

pe, pf, pv = dp.eval(data["coord"][0], data["box"][0], atype)
#pe, pf, pv = dp.eval(data["coord"].reshape(1,-1,3), data["box"][0].reshape(1,3,3), atype)

print('orignal energy:%f'%data['energy'][0])
print('predicted energy:%f'%pe)

output:
orignal energy:-1347.496799
predicted energy:-1322.281900

dp test -m graph.pb -s ../validation_data
output:
![image](https://github.com/deepmodeling/deepmd-kit/assets/100009402/981fb142-7802-43d7-bd24-14eaa131f990)

**Comment by njzjz** [comment]
See https://github.com/deepmodeling/deepmd-kit/discussions/3020#discussioncomment-7731948

**Reply by gust-07** [reply]
I have flowed your advice mentioned in this discussion. But still training data does not equal to predication data. Here is the result:

d_poscar = dpdata.System("../training_data", fmt = "deepmd/npy")
predict = d_poscar.predict(dp)
print("predicted energy:%s"%predict['energies'])

output:
predicted energy:[-1321.17637819 -1321.18011108 -1321.20414976 ... -1321.29880732
 -1321.54442515 -1321.6688769 ]

You said that the type map in the data and model cannot be assumed to be the same. How to deal with it. And I checked the d_poscar['atom_atype'] list, it was the same with my training type.raw. Did I misunderstand your words?

What's more, I hope to use the training model to do the flowing MD simulation in I-PI, so I choose python interface to do this job. Sadly, the results were bad, and that's why I retested this model and found this mismatch.

**Reply by njzjz** [reply]
The model type map can be obtained from `dp.get_type_map()`.

`dp test` is a wrapper of `DeepPot`, so when `dp test` gives the correct result, there should be nothing wrong with `DeepPot`.

**Reply by gust-07** [reply, marked as answer]
Thank you for your help. Problem solved. It seemed that **dp.eval** wouldn't differ different atom type with the same element name. So I retrained my model with the default atom types created by **dpdata**, and finally  done.
